File: 3607-power-grid-maintenance/3607-power-grid-maintenance.cpp
class Union {
    public:
        vector<int> parent;
        vector<int> rank;
        unordered_map<int,set<int>> online;

        Union(int n) {
            parent.resize(n);
            rank.resize(n);
            for(int i=0;i<n;i++) {
                parent[i]=i;
                set<int> st;
                st.insert(i);
                online[i] = st;
            }
        }

        int findParent(int n) {
            while(n!=parent[n]) n = parent[parent[n]];
            return n;
        }

        void merge(int n1,int n2) {
            int p1 = findParent(n1);
            int p2 = findParent(n2);
            if(p1==p2) return;
            if(rank[p1]>=rank[p2]) {
                parent[p2]=p1;
                rank[p1]++;
                online[p1].insert(online[p2].begin(),online[p2].end());
            } else {
                parent[p1]=p2;
                rank[p2]++;
                online[p2].insert(online[p1].begin(),online[p1].end());
            }
            return;
        }

        void turnOffline(int n) {
            int p = findParent(n);
            online[p].erase(n);
        }

        int performMaintenance(int n) {
            int p = findParent(n);
            if(online[p].empty()) return -1;
            if(online[p].count(n)) return n;
            return *online[p].begin();
        } 
};

class Solution {
public:
    vector<int> processQueries(int c, vector<vector<int>>& connections, vector<vector<int>>& queries) {
        Union root(c+1);
        for(int i=0;i<connections.size();i++) root.merge(connections[i][0],connections[i][1]);
        vector<int> ans;
        for(int i=0;i<queries.size();i++) {
            int node = queries[i][1];
            if(queries[i][0]==1) ans.push_back(root.performMaintenance(node));
            else root.turnOffline(node);
        }
        return ans;
    }
};


/*


DSU => iterate connections array and populate dsu
unordered_map<int,set<int>> => merge function contains logic for unioning the 2 sets into parent key in map based on rank



*/

File: 3607-power-grid-maintenance/3607-power-grid-maintenance.java
class Union {
    private int[] parent;
    private int[] rank;
    public Map<Integer,TreeSet<Integer>> online;
    public Union(int n) {
        parent = new int[n];
        rank = new int[n];
        online = new HashMap<>();
        for(int i=0;i<n;i++) { 
            TreeSet<Integer> st = new TreeSet<>();
            st.add(i);
            online.put(i,st);
            parent[i]=i;
        }
    }

    public int findParent(int n) {
        while(n!=parent[n]) n = parent[parent[n]];
        return n;
    }

    public boolean merge(int n1,int n2) {
        int p1 = findParent(n1);
        int p2 = findParent(n2);
        if(p1==p2) return true;
        if(rank[p1]>=rank[p2]) {
            parent[p2] =  p1;
            rank[p1]++;
            TreeSet<Integer> st = online.get(p1);
            st.addAll(online.get(p2));
        } else {
            parent[p1] = p2;
            rank[p2]++;
            TreeSet<Integer> st = online.get(p2);
            st.addAll(online.get(p1));
        }
        return true;
    }

    public void turnOffline(int n) {
        int p = findParent(n);
        online.get(p).remove(n);
        return;
    }

    public int performMaintanance(int n) {
        int p = findParent(n);
        TreeSet<Integer> st = online.get(p);
        if(st.size()==0) return -1;
        int ans = st.contains(n)?n:st.first();
        return ans; 
    }
};

class Solution {
    public int[] processQueries(int c, int[][] connections, int[][] queries) {
        Union root = new Union(c+1);
        for(int i=0;i<connections.length;i++) root.merge(connections[i][0],connections[i][1]);
        List<Integer> ans = new ArrayList<>();
        for(int i=0;i<queries.length;i++) {
            int node = queries[i][1];
            if(queries[i][0]==1) {
                ans.add(root.performMaintanance(node));  
            } else {
                root.turnOffline(node);
            }
        }
        return ans.stream().mapToInt(Integer::intValue).toArray();
    }
}